### PR TITLE
Updated GLUE script. New feature: Binary mask creation from the tokenizer's encoding.

### PR DESCRIPTION
**Pull request by LysandreJik**
The new `tokenizer.encode(seq_0, seq_1, add_special_tokens=True)` method makes life easier when building sequences. However, it makes it harder to create binary masks as the different sequence lengths are unknown. As a feature, I have therefore added a flag to the encode function so that it can output binary masks.

Example:
```py
from pytorch_transformers import BertTokenizer

tokenizer = BertTokenizer.from_pretrained("bert-base-cased")
seq_0 = "This is the one"
seq_1 = "This is the last"

input_ids, mask = tokenizer.encode(seq_0, seq_1, add_special_tokens=True, output_mask=True)
# input_ids: [ 101, 1188, 1110, 1103, 1141,  102, 1188, 1110, 1103, 1314, 102]
# mask:      [   0,    0,    0,    0,    0,    0,    1,    1,    1,    1,   1]
```

It works for BERT, RoBERTa, XLM, and XLNet. I have refactored the GLUE example with this method. It greatly simplifies input creation. 

I have added an additional unit test to the `commontests` suite. Furthermore, in order to make sure the tokenization was correct I compared against the original input creation of the GLUE script to make sure every encoded sequence remained the same.

### CODE REVIEW

**Comment by thomwolf** [comment]
I like it a lot! Could we also update `run_squad` similarly maybe?

File: pytorch_transformers/tokenization_utils.py [hunk 1]
```diff
@@ -696,15 +698,17 @@ def encode(self, text, text_pair=None, add_special_tokens=False, **kwargs):
...
+            if output_mask:
+                logger.warning("Can't output mask if no special tokens are involved. Please call the method with add_special_tokens set to True.")
             return first_sentence_tokens, second_sentence_tokens
 
     def add_special_tokens_single_sentence(self, token_ids):
         logger.warning("This tokenizer does not make use of special tokens. The sequence has been returned with no modification.")
         return token_ids
 
-    def add_special_tokens_sentences_pair(self, token_ids_0, token_ids_1):
+    def add_special_tokens_sentences_pair(self, token_ids_0, token_ids_1, output_mask=False):
```
Comment (thomwolf) [inline pytorch_transformers/tokenization_utils.py]: I think we should still output a mask here for consistency of the return type (tuple) across models. Maybe just a list of zeros.

File: pytorch_transformers/tokenization_utils.py [hunk 2]
```diff
@@ -696,15 +698,17 @@ def encode(self, text, text_pair=None, add_special_tokens=False, **kwargs):
         second_sentence_tokens = [self._convert_token_to_id(token) for token in self.tokenize(text_pair, **kwargs)]
 
         if add_special_tokens:
-            return self.add_special_tokens_sentences_pair(first_sentence_tokens, second_sentence_tokens)
+            return self.add_special_tokens_sentences_pair(first_sentence_tokens, second_sentence_tokens, output_mask)
         else:
+            if output_mask:
+                logger.warning("Can't output mask if no special tokens are involved. Please call the method with add_special_tokens set to True.")
```
Comment (thomwolf) [inline pytorch_transformers/tokenization_utils.py]: Do we really need `special_tokens` to have an output mask?

File: pytorch_transformers/tokenization_utils.py (hunk 1 above)
Comment (LysandreJik) [reply pytorch_transformers/tokenization_utils.py]: This would introduce a breaking change in the encoder method, wouldn't it?

File: pytorch_transformers/tokenization_utils.py (hunk 2 above)
Comment (LysandreJik) [reply pytorch_transformers/tokenization_utils.py]: Hmm this warning is put here so that if a user asks for an `output_mask` while not joining sequences (`add_special_tokens` is set to `False`), he is aware that no mask will be returned.

If the sequences are not joined together, I don't think it makes sense to output a mask,  does it?

Comment (julien-c) [reply pytorch_transformers/tokenization_utils.py]: Maybe the warning can be something like `"Can't output mask if you're not joining two sequences"`

**Comment by LysandreJik** [comment]
Only need to adapt the SQuAD script with the new encode w/mask as well as DistilBERT and I think we're good to go @julien-c

File: examples/utils_glue.py [hunk 3]
```diff
@@ -415,58 +405,14 @@ def convert_examples_to_features(examples, label_list, max_seq_length,
...
-            tokens = [cls_token] + tokens
-            segment_ids = [cls_token_segment_id] + segment_ids
-
-        input_ids = tokenizer.convert_tokens_to_ids(tokens)
+        inputs = tokenizer.encode_plus(
+            example.text_a,
+            example.text_b,
+            add_special_tokens=True,
+            output_mask=True,
+            max_length=max_seq_length
```
Comment (thomwolf) [inline examples/utils_glue.py]: Add the command that said the first sentence is to be truncated in priority (for people reading the examples as reference)

File: pytorch_transformers/tokenization_distilbert.py [hunk 4]
```diff
@@ -60,3 +60,21 @@ class DistilBertTokenizer(BertTokenizer):
     vocab_files_names = VOCAB_FILES_NAMES
     pretrained_vocab_files_map = PRETRAINED_VOCAB_FILES_MAP
     max_model_input_sizes = PRETRAINED_POSITIONAL_EMBEDDINGS_SIZES
+
+    def add_special_tokens_single_sequence(self, token_ids):
```
Comment (thomwolf) [inline pytorch_transformers/tokenization_distilbert.py]: Are sure about the format for encoding DistilBert sequences and pair of sequences (I thought it used a classifier token as well). To be confirmed with @VictorSanh

File: pytorch_transformers/tokenization_distilbert.py [hunk 5]
```diff
@@ -60,3 +60,21 @@ class DistilBertTokenizer(BertTokenizer):
...
+        return token_ids
+
+    def add_special_tokens_sequence_pair(self, token_ids_0, token_ids_1):
+        sep = [self.sep_token_id]
```
Comment (thomwolf) [inline pytorch_transformers/tokenization_distilbert.py]: To be confirmed with @VictorSanh

File: pytorch_transformers/tokenization_utils.py [hunk 6]
```diff
@@ -518,6 +518,36 @@ def add_tokens(self, new_tokens):
...
+
+        Returns:
+            Number of tokens added to sequences
+        """
+
+        if pair:
+            initial_tokens_len = len(self.encode("This is a sequence") + self.encode("This is another"))
+            final_tokens = self.encode("This is a sequence", "This is another", add_special_tokens=True)
+
+            # In some models (e.g. GPT-2), there is no sequence pair encoding.
```
Comment (thomwolf) [inline pytorch_transformers/tokenization_utils.py]: Encoding a pair of sequence with GPT2 is a silent error then?
Maybe we should raise an error or do something that makes basic sense (pure concatenation of token sequences).

File: pytorch_transformers/tokenization_utils.py [hunk 7]
```diff
@@ -518,6 +518,36 @@ def add_tokens(self, new_tokens):
...
+            if len(final_tokens) == 2:
```
Comment (thomwolf) [inline pytorch_transformers/tokenization_utils.py]: Because I don't like this `len(final_tokens) == 2` test which is strange and makes no sense for the user/reader of the library.

File: pytorch_transformers/tokenization_utils.py [hunk 8]
```diff
@@ -666,35 +696,146 @@ def _convert_token_to_id(self, token):
...
-        
+
         Same as doing ``self.convert_tokens_to_ids(self.tokenize(text))``.
 
         Args:
             text: The first sequence to be encoded.
             text_pair: Optional second sequence to be encoded.
             add_special_tokens: if set to ``True``, the sequences will be encoded with the special tokens relative
                 to their model.
-            **kwargs: passed to the `self.tokenize()` method
```
Comment (thomwolf) [inline pytorch_transformers/tokenization_utils.py]: kwargs is still passed to the `tokenize()` methode isn't it?

File: pytorch_transformers/tokenization_utils.py [hunk 9]
```diff
@@ -666,35 +696,146 @@ def _convert_token_to_id(self, token):
     def encode(self, text, text_pair=None, add_special_tokens=False, **kwargs):
         """
         Converts a string in a sequence of ids (integer), using the tokenizer and vocabulary.
-        
+
         Same as doing ``self.convert_tokens_to_ids(self.tokenize(text))``.
 
         Args:
             text: The first sequence to be encoded.
```
Comment (thomwolf) [inline pytorch_transformers/tokenization_utils.py]: We should explain clearly the various possible inputs of the method in the docstring

File: pytorch_transformers/tokenization_utils.py [hunk 10]
```diff
@@ -666,35 +696,146 @@ def _convert_token_to_id(self, token):
...
             text_pair: Optional second sequence to be encoded.
             add_special_tokens: if set to ``True``, the sequences will be encoded with the special tokens relative
                 to their model.
-            **kwargs: passed to the `self.tokenize()` method
         """
         if text_pair is None:
             if add_special_tokens:
-                return self.add_special_tokens_single_sentence(self.convert_tokens_to_ids(self.tokenize(text, **kwargs)))
+                sequence_tokens = self.convert_tokens_to_ids(self.tokenize(text, **kwargs)) if isinstance(text, six.string_types) else text
```
Comment (thomwolf) [inline pytorch_transformers/tokenization_utils.py]: Here:
- the `else text` option should be inside the call to `self.convert_tokens_to_ids()`
- we should add the option to supply a list of integer: another `else text` to bypass `self.convert_tokens_to_ids` with a test: `isinstance(text, (list, tuple)) and len(text) > 0 and isinstance(text[0], int)`

File: pytorch_transformers/tokenization_utils.py [hunk 11]
```diff
@@ -666,35 +696,146 @@ def _convert_token_to_id(self, token):
...
+                return self.add_special_tokens_single_sequence(sequence_tokens)
             else:
-                return self.convert_tokens_to_ids(self.tokenize(text, **kwargs))
+                ids = self.convert_tokens_to_ids(self.tokenize(text, **kwargs)) if isinstance(text, six.string_types) else text
```
Comment (thomwolf) [inline pytorch_transformers/tokenization_utils.py]: Same as above

File: pytorch_transformers/tokenization_utils.py [hunk 12]
```diff
@@ -666,35 +696,146 @@ def _convert_token_to_id(self, token):
...
+                return ids
 
-        first_sentence_tokens = [self._convert_token_to_id(token) for token in self.tokenize(text, **kwargs)]
-        second_sentence_tokens = [self._convert_token_to_id(token) for token in self.tokenize(text_pair, **kwargs)]
+        first_sentence_tokens = [self._convert_token_to_id(token) for token in self.tokenize(text, **kwargs)] if isinstance(text, six.string_types) else text
```
Comment (thomwolf) [inline pytorch_transformers/tokenization_utils.py]: same as above

File: pytorch_transformers/tokenization_utils.py [hunk 13]
```diff
@@ -666,35 +696,146 @@ def _convert_token_to_id(self, token):
...
+            return self.add_special_tokens_sequence_pair(first_sentence_tokens, second_sentence_tokens)
+        else:
+            logger.warning("No special tokens were added. The two sequences have been concatenated.")
+            return first_sentence_tokens + second_sentence_tokens
+
+    def encode_plus(self,
+                    text,
+                    text_pair=None,
+                    add_special_tokens=False,
+                    output_mask=False,
```
Comment (thomwolf) [inline pytorch_transformers/tokenization_utils.py]: I'm not a fan of using the work `mask` here which is similar to an attention mask but we are actually talking about what we call `token_type_ids` in the rest of the library. Maybe switch to `output_token_type` and change the vocabulary in the method as well.

File: pytorch_transformers/tokenization_utils.py [hunk 14]
```diff
@@ -666,35 +696,146 @@ def _convert_token_to_id(self, token):
...
+                    max_length=None,
+                    stride=0,
+                    truncate_second_sequence_first=True,
```
Comment (thomwolf) [inline pytorch_transformers/tokenization_utils.py]: I think we want the default to be `truncate_first_sequence=True`. Also thinking about it again, I think it may no sense to talk about priority as we don't have a "threshold" after which the other sequence will be truncated. Let's just make it a `truncate_second_sequence=False` that we can set to `True` to truncate the second sequence instead of the first one.

File: pytorch_transformers/tokenization_utils.py [hunk 15]
```diff
@@ -666,35 +696,146 @@ def _convert_token_to_id(self, token):
     def encode(self, text, text_pair=None, add_special_tokens=False, **kwargs):
```
Comment (thomwolf) [inline pytorch_transformers/tokenization_utils.py]: Didn't we say that `encode` should now just be a front end to `encode_plus` filtering the output?

File: pytorch_transformers/tokenization_utils.py [hunk 16]
```diff
@@ -666,35 +696,146 @@ def _convert_token_to_id(self, token):
...
+                from the main sequence returned. The value of this argument defined the number of additional tokens.
+            truncate_second_sequence_first: if there is a specified max_length, this flag will choose which sequence
+                will be truncated.
+            **kwargs: passed to the `self.tokenize()` method
+        """
+
+        information = {}
+
+        if text_pair is None:
+            sequence_tokens = self.convert_tokens_to_ids(self.tokenize(text, **kwargs)) if isinstance(text, six.string_types) else text
```
Comment (thomwolf) [inline pytorch_transformers/tokenization_utils.py]: All the remarks I made for `encode` above apply here as well.

File: pytorch_transformers/tokenization_utils.py [hunk 17]
```diff
@@ -666,35 +696,146 @@ def _convert_token_to_id(self, token):
...
+            if add_special_tokens:
+                information = self.prepare_for_model(sequence_tokens, max_length, stride)
```
Comment (thomwolf) [inline pytorch_transformers/tokenization_utils.py]: call `prepare_for_model` with keywords instead of positional arguments as this method will likely see its signature change in the future with additional options.

File: pytorch_transformers/tokenization_utils.py (hunk 6 above)
Comment (LysandreJik) [reply pytorch_transformers/tokenization_utils.py]: Encoding a pair of sequences with GPT-2 concatenates them together and prints the following warning: `This tokenizer does not make use of special tokens. The two sequences have been concatenated.`

File: pytorch_transformers/tokenization_utils.py (hunk 7 above)
Comment (LysandreJik) [reply pytorch_transformers/tokenization_utils.py]: You're totally right, this indeed does not make sense. It is an artifact from the previous `encode` function on which I was working. Since the tokens are now concatenated in models that do not handle sequence pairs, this test will never pass. I'm removing it.

File: pytorch_transformers/tokenization_utils.py (hunk 8 above)
Comment (LysandreJik) [reply pytorch_transformers/tokenization_utils.py]: You're right, it shouldn't have been removed.

File: pytorch_transformers/tokenization_utils.py (hunk 10 above)
Comment (LysandreJik) [reply pytorch_transformers/tokenization_utils.py]: I was only taking into account the list of integers (token ids), not a list of tokens. Supplying the three formats: `string`, `list[string]` and `list[int]` makes sense, I'm changing this now.

File: pytorch_transformers/tokenization_utils.py (hunk 15 above)
Comment (LysandreJik) [reply pytorch_transformers/tokenization_utils.py]: You're right, it's cleaner this way

File: pytorch_transformers/tokenization_utils.py (hunk 13 above)
Comment (LysandreJik) [reply pytorch_transformers/tokenization_utils.py]: Switched! Thanks.

File: pytorch_transformers/tokenization_utils.py (hunk 14 above)
Comment (LysandreJik) [reply pytorch_transformers/tokenization_utils.py]: I agree, this was misleading.

File: pytorch_transformers/tokenization_utils.py [hunk 18]
```diff
@@ -666,35 +690,188 @@ def _convert_token_to_id(self, token):
...
+
         if text_pair is None:
+            sequence_tokens = get_input_ids(text)
+
             if add_special_tokens:
-                return self.add_special_tokens_single_sentence(self.convert_tokens_to_ids(self.tokenize(text, **kwargs)))
+                information = self.prepare_for_model(sequence_tokens, max_length=max_length, stride=stride)
             else:
-                return self.convert_tokens_to_ids(self.tokenize(text, **kwargs))
+                if max_length:
```
Comment (thomwolf) [inline pytorch_transformers/tokenization_utils.py]: Shouldn't we have this `max_length` test before the `add_special_tokens` test (and one after the other instead of being in a `if..then` structure?

File: pytorch_transformers/tokenization_utils.py [hunk 19]
```diff
@@ -666,35 +690,188 @@ def _convert_token_to_id(self, token):
...
+                    stride=stride
+                )
+
+                if output_token_type:
+                    information["token_type_ids"] = self.create_token_type_ids_from_sequences(text, text_pair)
+            else:
+                logger.warning("No special tokens were added. The two sequences have been concatenated.")
+                sequence = first_sentence_tokens + second_sentence_tokens
+
+                if max_length:
```
Comment (thomwolf) [inline pytorch_transformers/tokenization_utils.py]: Same here

Comment (thomwolf) [reply pytorch_transformers/tokenization_utils.py]: Oh yes I understand actually

File: pytorch_transformers/tokenization_utils.py (hunk 18 above)
Comment (thomwolf) [reply pytorch_transformers/tokenization_utils.py]: Ok, see that's in the `prepare_model` methode
